src/magicbox/ResultPreviewsManager.ts

```diff
@@ -0,0 +1,214 @@
+  previewsQuery: Promise<ISearchResultPreview[]>;
+}
+
+export enum ResultPreviewsManagerEvents {
+  PopulateSearchResultPreviews = 'populateSearchResultPreviews'
+}
+
+export interface ISearchResultPreview {
+  element: HTMLElement;
+  onSelect: () => void;
+}
+
+export interface IResultPreviewsManagerOptions {
+  previewClass: string;
+  selectedClass: string;
+  previewHeaderText: string;
+}
+
+export class ResultPreviewsManager {
+  private options: IResultPreviewsManagerOptions;
+  private suggestionsPreviewContainer: Dom;
+  private resultPreviewsHeader: Dom;
+  private resultPreviewsContainer: Dom;
+  private lastPreviewsQuery: Promise<ISearchResultPreview[]>;
+  private lastSelectedSuggestion: HTMLElement;
+  private root: HTMLElement;
+
+  public get previewsOwner() {
+    return this.lastSelectedSuggestion;
+  }
+
+  public get hasPreviews() {
+    return !!this.suggestionsPreviewContainer;
+  }
+
+  public get focusedPreviewElement() {
+    if (!this.hasPreviews) {
+      return null;
+    }
+    const focusedElement = this.suggestionsPreviewContainer.findClass(this.options.selectedClass)[0];
+    if (!focusedElement || !focusedElement.classList.contains(this.options.previewClass)) {
+      return null;
+    }
+    return focusedElement;
+  }
+
+  public get previewElements() {
+    if (!this.hasPreviews) {
+      return [];
+    }
+    return this.suggestionsPreviewContainer.findClass(this.options.previewClass);
+  }
+
+  private get suggestionsListbox() {
+    return $$($$(this.element).findClass('coveo-magicbox-suggestions')[0]);
+  }
+
+  private get numberOfResultsPerRow() {
+    const previewSelectables = this.suggestionsPreviewContainer.findClass(this.options.previewClass);
+    if (previewSelectables.length === 0) {
+      return 0;
+    }
+    const firstVerticalOffset = previewSelectables[0].offsetTop;
+    const firstIndexOnNextRow = findIndex(previewSelectables, previewSelectable => previewSelectable.offsetTop !== firstVerticalOffset);
+    return firstIndexOnNextRow !== -1 ? firstIndexOnNextRow : previewSelectables.length;
+  }
+
+  constructor(private element: HTMLElement, options: Partial<IResultPreviewsManagerOptions> = {}) {
+    this.options = defaults(options, <IResultPreviewsManagerOptions>{
+      previewHeaderText: l('QuerySuggestPreview'),
+      previewClass: 'coveo-preview-selectable',
+      selectedClass: 'magic-box-selected'
+    });
+    this.root = Component.resolveRoot(element);
+  }
+
+  public async displaySearchResultPreviewsForSuggestion(suggestion: HTMLElement) {
+    if (this.lastSelectedSuggestion !== suggestion) {
+      await this.loadSearchResultPreviews(suggestion);
+    }
+  }
+
+  public getElementInDirection(direction: Direction) {
+    const previewElements = this.previewElements;
+    const focusedIndex = previewElements.indexOf(this.focusedPreviewElement);
+
+    if (focusedIndex === -1) {
+      return null;
+    }
+
+    if (focusedIndex === 0 && direction === Direction.Left) {
```

Comment:
`focusedIndex === 0` could be changed to `focusedIndex % previewElements.length === 0` if we want the left keyboard key to go back to suggestions from any row instead of only the first row.